java/src/main/java/ai/rapids/cudf/ParquetColumnWriterOptions.java

```diff
@@ -0,0 +1,400 @@
+    }
+
+    /**
+     * Set nullable column meta data
+     */
+    public T withNullableColumn(String... name) {
+      withColumn(true, name);
+      return (T) this;
+    }
+
+    /**
+     * Set a simple child meta data
+     * @return this for chaining.
+     */
+    public T withColumn(boolean nullable, String... name) {
+      for (String n : name) {
+        children.add(withColumn(n, nullable));
+      }
+      return (T) this;
+    }
+
+    /**
+     * Set a Decimal child meta data
+     * @return this for chaining.
+     */
+    public T withDecimalColumn(String name, int precision, boolean nullable) {
+      children.add(withDecimal(name, precision, nullable));
+      return (T) this;
+    }
+
+    /**
+     * Set a Decimal child meta data
+     * @return this for chaining.
+     */
+    public T withNullableDecimalColumn(String name, int precision) {
+      withDecimalColumn(name, precision, true);
+      return (T) this;
+    }
+
+    /**
+     * Set a Decimal child meta data
+     * @return this for chaining.
+     */
+    public T withDecimalColumn(String name, int precision) {
+      withDecimalColumn(name, precision, false);
+      return (T) this;
+    }
+
+    /**
+     * Set a timestamp child meta data
+     * @return this for chaining.
+     */
+    public T withTimestampColumn(String name, boolean isInt96, boolean nullable) {
+      children.add(withTimestamp(name, isInt96, nullable));
+      return (T) this;
+    }
+
+    /**
+     * Set a timestamp child meta data
+     * @return this for chaining.
+     */
+    public T withTimestampColumn(String name, boolean isInt96) {
+      withTimestampColumn(name, isInt96, false);
+      return (T) this;
+    }
+
+    /**
+     * Set a timestamp child meta data
+     * @return this for chaining.
+     */
+    public T withNullableTimestampColumn(String name, boolean isInt96) {
+      withTimestampColumn(name, isInt96, true);
+      return (T) this;
+    }
+
+    public abstract V build();
+  }
+
+  ParquetColumnWriterOptions(String columnName, boolean isTimestampTypeInt96,
+                             int precision, boolean isNullable) {
+    this.isTimestampTypeInt96 = isTimestampTypeInt96;
+    this.precision = precision;
+    this.isNullable = isNullable;
+    this.columName = columnName;
+  }
+
+  ParquetColumnWriterOptions(String columnName, boolean isNullable) {
+    this.isTimestampTypeInt96 = false;
+    this.precision = 0;
+    this.isNullable = isNullable;
+    this.columName = columnName;
+  }
+
+  ParquetColumnWriterOptions(String columnName) {
+    this(columnName, true);
+  }
+
+  boolean[] getFlatIsTimeTypeInt96() {
+    boolean[] ret = {isTimestampTypeInt96};
+
+    for (ParquetColumnWriterOptions opt: childColumnOptions) {
+      boolean[] b = opt.getFlatIsTimeTypeInt96();
+      boolean[] tmp = new boolean[ret.length + b.length];
+      System.arraycopy(ret, 0, tmp, 0, ret.length);
+      System.arraycopy(b, 0, tmp, ret.length, b.length);
+      ret = tmp;
```

Comment:
This ends up creating a new temporary array of the result for each child and at every level.  Gathering all the child column name results and then concatenate them together at the end, when the final array size for this level can be computed directly, will generate less garbage and reduce the number of array copies.  This comment also applies to the other places where this algorithm was replicated.